logstash-core/src/main/java/org/logstash/plugin/Input.java

```diff
@@ -0,0 +1,44 @@
+package org.logstash.plugin;
+
+import org.logstash.Event;
+
+import java.util.Collection;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executors;
+import java.util.function.Consumer;
+
+/**
+ * A Logstash input plugin.
+ * <p>
+ * Input plugins produce events intended given to Processors.
+ * <p>
+ * Inputs will generally run forever until there is some terminating condition such as a pipeline shutdown.
+ */
+public interface Input {
+    /**
+     * The method contract is as follows:
+     * <p>
+     * 1) When `Thread.interrupted()` is true, you MUST close all open resources and return.
+     * 2) When this method returns, it is assumed all work for this input is completed.
+     * 3) consumer.accept(...) may block
+     * <p>
+     * Acknowledging Data:
+     * <p>
+     * When `consumer.accept(...)` returns, the given Events have been successfully written into the consumer, and it is now
+     * safe to acknowledge these events to the upstream data source. The consumer is generally the Logstash internal queue,
+     * such as the persistent queue.
+     *
+     * @param consumer Send batches of events with consumer.accept(batch).
+     */
+    void run(Consumer<Collection<Event>> consumer);
+
+    /**
+     * Return this Input as a Callable for use with an ExecutorService.
+     *
+     * @param consumer the
+     * @return A Callable which wraps the `run` method.
+     */
+    default Callable<Void> toCallable(Consumer<Collection<Event>> consumer) {
```

Comment:
@jordansissel I don't think it makes all that much sense to allow anything but single `Event` here with our current ack strategy.
How will we enforce something like the max unread count if inputs can pass arbitrarily sized `Collection`s to the Queue? (come to think of it, this is a problem in the current implementation as well ... probably better not to perpetuate this issue and only accept single `Event` here).
This also allows for much nicer batching to an IO backed queue, you can simply collect up to `n` `Event`  to an `ArrayBlockingQueue` and batch write that thing if you feel like you have enough elements to make a batch write worthwhile.
On the other hand you don't force the in memory Queue to handle the overhead of receiving a batch (it's really just overhead for an `ArrayBlockingQueue` because it can't enqueue a batch and it's kind of hard to make a fast concurrent Queue that has batched writes without allocations ... just enqueuing `Collection<Event>` to an`ArrayBlockingQueue` seems like a really bad idea here in terms of performance and also memory usage under pressure).